java/arcs/core/host/ArcHostContextParticle.kt

```diff
@@ -0,0 +1,255 @@
+            throw IllegalStateException("Unable to serialize $arcId for $hostId", e)
+        }
+    }
+
+    /**
+     * Reads [ArcHostContext] from serialized representation as Arcs Schema types. See
+     * `ArcHostContext.arcs' for Schema definitions. NOTE: This is more complex than it needs
+     * to be because references are not supported yet in schema2kotlin, and so this information
+     * is stored in de-normalized format.
+     */
+    suspend fun readArcHostContext(
+        arcId: String,
+        hostId: String,
+        instantiateParticle: suspend (ParticleIdentifier) -> Particle
+    ): ArcHostContext? {
+        try {
+            val arcStateEntity = handles.arcHostContext.fetch()
+            val particleEntities = handles.particles.fetchAll()
+            val connectionEntities = handles.handleConnections.fetchAll()
+
+            // This arc has never been serialized before
+            if (arcStateEntity == null ||
+                particleEntities.isEmpty() ||
+                connectionEntities.isEmpty()) {
+                return null
+            }
+
+            // construct a map of particleName to Particle instance
+            val instantiatedParticles = particleEntities.map {
+                it.particleName to instantiateParticle(ParticleIdentifier.from(it.location))
+            }.associateBy({ it.first }, { it.second })
+
+            // construct a map from particleName to a list of handleNames mapped to HandleConnection
+            val handleConnections = connectionEntities.map { entity ->
+                entity.particleName to createHandleConnection(
+                    entity,
+                    arcId,
+                    hostId,
+                    instantiatedParticles
+                )
+            }.groupBy { it.first }
+
+            // construct a map from Plan.Particle to ParticleContext
+            val particles = particleEntities.map {
+                Plan.Particle(
+                    it.particleName,
+                    it.location,
+                    createHandlesMap(handleConnections, it, arcId)
+                ) to ParticleContext(
+                    requireNotNull(instantiatedParticles[it.particleName]) {
+                        "${it.particleName} not instantiable for $arcId and $hostId"
+                    },
+                    mutableMapOf(),
+                    ParticleState.valueOf(it.particleState),
+                    it.consecutiveFailures.toInt()
+                )
+            }.associateBy({ it.first }, { it.second })
+
+            return ArcHostContext(particles.toMutableMap(), arcStateEntity?.let {
+                ArcState.valueOf(it.arcState)
+            } ?: ArcState.NeverStarted)
+        } catch (e: Exception) {
+            throw IllegalStateException("Unable to deserialize $arcId for $hostId", e)
+        }
+    }
+
+    private fun createHandlesMap(
+        handleConnections: Map<String, List<Pair<String, Pair<String, Plan.HandleConnection>>>>,
```

Comment:
This feels way too complicated. Can this giant map of collections be broken down into a nicer data class or a map of String->some data class)?